### DISCUSSION TITLE
How to extract the concentration time series for two pollutants using PySWMM?

### DISCUSSION THREAD

**Discussion by Freestyleyang** [pyswmm/pyswmm#514] [Q&A] [answered]
Hi There,

I defined two pollutants "TP" and "TSS" for a SWMM simulation. Using "**POLLUT_CONC_0**", I can extract the concentration series for "TP", **however, the concentration series for "TSS" is not returned**. I wonder how to extract the concentration time series for **both** pollutants? Thank you.

`from swmm.toolkit.shared_enum import NodeAttribute`
`from pyswmm import Simulation, Nodes, Output`

`with Output('Example1.out') as out:`
    `a = out.node_series("18", NodeAttribute.POLLUT_CONC_0, event_start[0], event_end[0])`

![01](https://github.com/pyswmm/pyswmm/assets/20953347/38e09cff-f1a7-44e6-89b7-01c96372ab18)

**Comment by bemcdonnell** [comment]
@Freestyleyang I am working on a simple solution for this.  Can you do me a favor and share your INP/OUT files here in a zip file? You can also email them directly to me.  I think i have both a short term fix (and will be working on the longer-term solution possibly for PySWMM-v2.

**Comment by Freestyleyang** [comment]
@bemcdonnell Thank you for your reply. Attached is the file for your reference. Would you please share your solution here? I am looking forward to the long-term solution! Thanks. 
[For Calibration.zip](https://github.com/pyswmm/pyswmm/files/14078809/For.Calibration.zip)

**Comment by bemcdonnell** [comment, marked as answer]
@Freestyleyang For now just code this enum into your script

```python
from aenum import Enum

class NodeAttribute(Enum, start = 0):
    """Node Attribute enum class.

    .. rubric:: Enum Members

    ========================= ===============================
    :attr:`~INVERT_DEPTH`     water depth above invert
    :attr:`~HYDRAULIC_HEAD`   hydraulic head
    :attr:`~PONDED_VOLUME`    volume stored and ponded
    :attr:`~LATERAL_INFLOW`   lateral inflow rate
    :attr:`~TOTAL_INFLOW`     total inflow rate
    :attr:`~FLOODING_LOSSES`  overflow rate
    :attr:`~POLLUT_CONC_0`    concentration of each pollutant
    ========================= ===============================
    """
    INVERT_DEPTH = 0
    HYDRAULIC_HEAD = 1
    PONDED_VOLUME = 2
    LATERAL_INFLOW = 3
    TOTAL_INFLOW = 4
    FLOODING_LOSSES = 5
    POLLUT_CONC_0 = 6
    POLLUT_CONC_1 = 7
```

You can add as many values as you have pollutants.  I will look into making this more dynamic later.

**Reply by bemcdonnell** [reply]
@Freestyleyang let me know how this idea works. It looks like i posted this right when you shared your model so i didn't test it.

**Comment by Freestyleyang** [comment]
@bemcdonnell I just tested your code and it worked well! Thank you for your help. Is there any expected date that you will implement this fix in PySWMM-v2?

**Reply by bemcdonnell** [reply]
I'll have to do some research on this a little but more.  The problem is is that for every model we could have a different number of pollutants.  The Enum class is currently set up to describe a "general" layout for an OUT file (not specific to each OUT file).  So if we open up `POLLUT_CONC_1 = 7`, we might run into some strange issues if those polluts do not exist.  Here are some improvements that we made to pyswmm-v2 already to make the OUT file more pythonic. The user doesn't have to deal with the Enums anymore.

https://github.com/pyswmm/pyswmm/pull/511

Let me talk to @karosc and @abhiramm7 to get their ideas on how we can accommodate these extra properties dynamically.  This might be left out for pyswmm-v2 (but could be in pyswmm-v2.1). Depends on the priority.  We are trying to get pyswmm-v2 out ASAP since it is a few weeks behind.

**Reply by Freestyleyang** [reply]
@bemcdonnell Thank you very much for your explanation and your hard work on PySWMM! I am looking forward to PySWMM-v2!

**Reply by karosc** [reply]
@bemcdonnell , [this is how I handle it](https://github.com/karosc/swmm-pandas/blob/main/src/swmm/pandas/output/output.py#L436-L451).....there are a few gotchyas though that I know about:

1. Pollutant names in SMWM can overlap with other parameter names. e.g. I could name my pollutant "_rainfall_" which would collide with the pre-existing subcatchment parameter "_rainfall_". [This is how I handle that issue](https://github.com/karosc/swmm-pandas/blob/main/src/swmm/pandas/output/output.py#L374-L408).
2. Extending the global enums in swmm.toolkit could have unintended consquences if you open more than one output file in a single python session that have diffenent pollutant name schemes. [This is how I solve that problem](https://github.com/karosc/swmm-pandas/blob/main/src/swmm/pandas/output/output.py#L141-L198).

**Reply by bemcdonnell** [reply]
@karosc i think we can build off of the new OUT series classes with your approach. Let’s integrate this into pyswmm-v2.1. Maybe we could even guide @Freestyleyang into helping us out with this since this user will reap the benefits the most? @Freestyleyang does that sound interesting to you?

**Reply by sausagechen** [reply]
> I'll have to do some research on this a little but more. The problem is is that for every model we could have a different number of pollutants. The Enum class is currently set up to describe a "general" layout for an OUT file (not specific to each OUT file). So if we open up `POLLUT_CONC_1 = 7`, we might run into some strange issues if those polluts do not exist. Here are some improvements that we made to pyswmm-v2 already to make the OUT file more pythonic. The user doesn't have to deal with the Enums anymore.
> 
> #511
> 
> Let me talk to @karosc and @abhiramm7 to get their ideas on how we can accommodate these extra properties dynamically. This might be left out for pyswmm-v2 (but could be in pyswmm-v2.1). Depends on the priority. We are trying to get pyswmm-v2 out ASAP since it is a few weeks behind.

Sorry to carry out this issue again, has v2.1 been available to output more than one pollutants?

**Reply by karosc** [reply]
@sausagechen, it has not yet been fixed. The implementation I described above it still in [swmm-pandas](https://github.com/karosc/swmm-pandas) though if you want to give that a try.

**Comment by Freestyleyang** [comment]
@bemcdonnell I'd love to contribute to "PySWMM"! Please let me know how I can help, thanks.
